src/CMake/VisItMacros.cmake

```diff
@@ -263,3 +263,76 @@ function(ADD_CMAKE_GEN_TARGET gen_name
 endfunction()
 
 
+##############################################################################
+# This macro appends to a CACHE var list denoted by the NAME argument.
+# Caller is responsible for unsetting the CACHE var when no longer needed
+# to avoid polluting the CACHE.
+#
+# Designed mainly for targets whose sources live in subdirectories,
+# as a means for them to add to the parent's list of source/includes/etc
+# to prepare for a 'blt_add_library' call which requires SOURCES.
+#
+##############################################################################
+
+macro(visit_append_list)
+    cmake_parse_arguments(arg "" "NAME" "ITEMS" ${ARGN})
+    if(NOT DEFINED arg_NAME OR NOT DEFINED arg_ITEMS)
+        message(FATAL_ERROR "visit_append_list called with invalid arguments. Must supply 'NAME' (name of list) and 'ITEMS' (list of items to be added to named list.)")
+    endif()
+    set(${arg_NAME} ${${arg_NAME}} ${arg_ITEMS} CACHE STRING "" FORCE)
+endmacro()
+
+##############################################################################
+# Adds a library target. Wrapper around blt_add_library so that CACHE vars
+# possibly created by visit_append_list for forming SOURCES/INCLUDES, etc
+# can be unset.
+#
+# ARGUMENTS:
+#    NAME         library name               REQUIRED
+#    SOURCES      [source1 [source2 ...]]    REQUIRED
+#    INCLUDES     [dir1 [dir2 ...]]          OPTIONAL
+#    DEFINES      [define1 [define2 ...]]    OPTIONAL
+#    DEPENDS      [dep1 ...]                 OPTIONAL
+#    OUTPUT_NAME  [name]                     OPTIONAL
+#    FEATURES     [feat1 [feat2 ...]]        OPTIONAL
+#    FOLDER       [name])                    OPTIONAL
+#
+##############################################################################
+
+macro(visit_add_library)
+    set(options)
+    set(singleValueArgs NAME OUTPUT_NAME FOLDER)
+    set(multiValueArgs SOURCES INCLUDES DEFINES DEPENDS FEATURES)
+
+    # parse the arguments
+    cmake_parse_arguments(arg
+        "${options}" "${singleValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    # Sanity checks
+    if(NOT arg_NAME)
+        message(FATAL_ERROR "visit_add_library() must be called with argument NAME <name>")
+    endif()
+
+    if (NOT arg_SOURCES)
+        message(FATAL_ERROR "visit_add_library(NAME ${arg_NAME} ...) called with no given sources")
+    endif()
+
+    blt_add_library(
+        NAME       ${arg_NAME}
+        SOURCES    ${arg_SOURCES}
+        INCLUDES   ${arg_INCLUDES}
+        DEFINES    ${arg_DEFINES}
+        DEPENDS_ON ${arg_DEPENDS}
+        FOLDER     ${arg_FOLDER})
+
+    if (arg_FEATURES)
+        target_compile_features(${arg_NAME} PRIVATE ${arg_FEATURES})
+    endif()
+    # vars that may have been created by calls to visit_append_list
+    unset(${arg_NAME}_SOURCES CACHE)
+    unset(${arg_NAME}_INCLUDES CACHE)
+    unset(${arg_NAME}_DEFINES CACHE)
+    unset(${arg_NAME}_DEPENDS CACHE)
+    unset(${arg_NAME}_FEATURES CACHE)
```

Comment:
Ran across another problem with the use of CACHE vars for the lists and having to unset them.
When we build serial and parallel versions of a library, the SOURCES and INCLUDES lists are the same, so it makes sense to only create one list and use a name  without the _ser/_par suffices of the target name. For instance, avtpipeline_SOURCES instead of avtpipeline_ser_SOURCES.

But then, the unset calls used here won't work.
I could add logic to remove _ser/_par from the target name before making the unset calls so that they would work.
That creates another problem in that after creating the serial library, all the lists would be unset, and the call to add the parallel library would fail.

Maybe I should abandon use of `blt_add_library`, and the  need to create these lists in order to satisfy its need to supply SOURCES with the call.
Keep the current use of CMake's `add_library` call without sources, and subdirs add their sources via `target_sources` call. 
Could then use `blt_register_library` if we really need blt to keep track of our library targets.

Reply:
Hmm...I don't like abandoning BLT functionality here. Circling back to your comment about CMake's `add_library` and `target_sources` functionality vs. BLT's `blt_add_library`, I think BLT may have been going for a simplification of interface. But, that siimplification forces one to build this list before `blt_add_library` can be called.

I read in the BLT documentation...

> This macro creates a true CMake target that can be altered by other CMake commands like normal, such as set_target_property().

Does this mean you can maybe use `blt_add_library` with essentially empty lists and use CMake's `target_sources` to build out those lists from subdirs as you would have via straight CMake?

Reply:
> Does this mean you can maybe use `blt_add_library` with essentially empty lists and use CMake's `target_sources` to build out those lists from subdirs as you would have via straight CMake?

I tried passing an empty SOURCES list and it bombs.

Reply:
So for two targets derived essentially from the same sources just compiled with different flags, is it really a bad thing to wind up with two instances of cache variables representing the lists of sources names for those targets?

Reply:
Sorry, at the moment, feel like I am already jumping through hoops to use blt, creating extra lists containing the same items seems overkill and unnecessary work.

Reply:
I think creating the second list might be doable, without having to actually list every source file again by creating a copy for the par libraries, eg before the call to `visit_add_library` for the serial library:
set(somelib_SOURCES_PAR ${somelib_SOURCES})
Then use the _PAR lists in the call to add the parallel library.